TestRefreshAndPurge flaky in racetest

=== RUN   TestRefreshAndPurge
==================
WARNING: DATA RACE
Write at 0x00c4203bcf40 by goroutine 359:
  istio.io/istio/mixer/adapter/list.TestRefreshAndPurge()
      /go/src/istio.io/istio/mixer/adapter/list/list_test.go:512 +0x4f2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous write at 0x00c4203bcf40 by goroutine 1001:
  istio.io/istio/mixer/adapter/list.TestRefreshAndPurge.func1()
      /go/src/istio.io/istio/mixer/adapter/list/list_test.go:500 +0x1cf
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:1947 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2694 +0xb9
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1830 +0x7dc

Goroutine 359 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:68 +0x22a

Goroutine 1001 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:2795 +0x364
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:280 +0xa2
==================
--- FAIL: TestRefreshAndPurge (0.02s)

--- a/mixer/adapter/list/list.go
+++ b/mixer/adapter/list/list.go
@@ -245,6 +245,14 @@ func (h *handler) purgeList() {
 	h.lock.Unlock()
 }
 
+func (h *handler) isPurged() bool {
+	h.lock.Lock()
+	result := h.list == nil
+	h.lock.Unlock()
+
+	return result
+}
+
 ///////////////// Bootstrap ///////////////
 
 // GetInfo returns the Info associated with this adapter implementation.

--- a/mixer/adapter/list/list.go
+++ b/mixer/adapter/list/list.go
@@ -245,9 +245,9 @@ func (h *handler) purgeList() {
 	h.lock.Unlock()
 }
 
-func (h *handler) isPurged() bool {
+func (h *handler) hasData() bool {
 	h.lock.Lock()
-	result := h.list == nil
+	result := h.list != nil
 	h.lock.Unlock()
 
 	return result

--- a/mixer/adapter/list/list_test.go
+++ b/mixer/adapter/list/list_test.go
@@ -21,13 +21,12 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"strconv"
-	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
 
 	"github.com/ghodss/yaml"
-	rpc "github.com/gogo/googleapis/google/rpc"
+	"github.com/gogo/googleapis/google/rpc"
 
 	"istio.io/istio/mixer/adapter/list/config"
 	"istio.io/istio/mixer/pkg/adapter/test"
@@ -472,32 +471,17 @@ func TestRegexErrors(t *testing.T) {
 }
 
 func TestRefreshAndPurge(t *testing.T) {
-	listToServe := "ABC"
-	var h *handler
-	var err error
-
-	wg := sync.WaitGroup{}
-	wg.Add(1)
 	var failServe int32
-	purgeDetected := false
 
+	// web server that returns valid data, until failServe is incremented at which point it
+	// returns errors
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-
 		if atomic.LoadInt32(&failServe) > 0 {
 			w.WriteHeader(http.StatusMovedPermanently)
-
-			if h.list == nil {
-				// the list was purged
-				if !purgeDetected {
-					purgeDetected = true
-					wg.Done()
-				}
-			}
-
 			return
 		}
 
-		if _, err = w.Write([]byte(listToServe)); err != nil {
+		if _, err := w.Write([]byte("ABC")); err != nil {
 			t.Errorf("w.Write failed: %v", err)
 		}
 	}))
@@ -509,11 +493,13 @@ func TestRefreshAndPurge(t *testing.T) {
 		Ttl:             2 * time.Millisecond,
 		EntryType:       config.STRINGS,
 	}
-	h, err = buildHandler(t, &cfg)
+
+	h, err := buildHandler(t, &cfg)
 	if err != nil {
 		t.Fatalf("Got error %v, expecting success", err)
 	}
 
+	// make sure everything is working OK
 	_, err = h.HandleListEntry(context.Background(), &listentry.Instance{Value: "ABC"})
 	if err != nil {
 		t.Errorf("Got error %v, expecting success", err)
@@ -523,8 +509,15 @@ func TestRefreshAndPurge(t *testing.T) {
 	atomic.AddInt32(&failServe, 1)
 
 	// wait for the list to have been purged
-	wg.Wait()
+	for {
+		time.Sleep(1 * time.Millisecond)
+		if h.isPurged() {
+			// list has been purged
+			break
+		}
+	}
 
+	// ensure we get an error now that the list has been purged
 	_, err = h.HandleListEntry(context.Background(), &listentry.Instance{Value: "ABC"})
 	if err == nil {
 		t.Error("Got success, expected error")

--- a/mixer/adapter/list/list_test.go
+++ b/mixer/adapter/list/list_test.go
@@ -499,6 +499,15 @@ func TestRefreshAndPurge(t *testing.T) {
 		t.Fatalf("Got error %v, expecting success", err)
 	}
 
+	// wait for the list to have been populated
+	for {
+		time.Sleep(1 * time.Millisecond)
+		if h.hasData() {
+			// list has been populated
+			break
+		}
+	}
+
 	// make sure everything is working OK
 	_, err = h.HandleListEntry(context.Background(), &listentry.Instance{Value: "ABC"})
 	if err != nil {
@@ -511,7 +520,7 @@ func TestRefreshAndPurge(t *testing.T) {
 	// wait for the list to have been purged
 	for {
 		time.Sleep(1 * time.Millisecond)
-		if h.isPurged() {
+		if !h.hasData() {
 			// list has been purged
 			break
 		}

